### PR TITLE
support for client-specific notifications in a multi-server set-up

### PR DESCRIPTION
**Pull request by jrochel**
this patch is not yet well-reviewed/tested by myself; but I'll will be doing this in rewriting Os_notif to be based Eliom_notif.

### CODE REVIEW

File: src/lib/eliom_notif.server.ml [hunk 1]
```diff
@@ -3,17 +3,41 @@ open Lwt
...
+  type server_notif
+  type client_notif
+  val listen : key -> unit
+  val unlisten : key -> unit
+  val notify : ?notfor:[`Me | `Id of identity] -> key -> server_notif -> unit
+  val client_ev : unit -> (key * client_notif) Eliom_react.Down.t Lwt.t
+  val clean : unit -> unit Lwt.t
+end
+
+module type MAKE = sig
```
Comment (vasilisp) [inline src/lib/eliom_notif.server.ml]: `MAKE` sounds a bit confusing for the argument to `Make`. Maybe `ARG`?

File: src/lib/eliom_notif.server.ml [hunk 2]
```diff
@@ -188,3 +217,29 @@ module Make (A : S) = struct
...
+  type notification
+  val equal_key                  : key -> key -> bool
+  val equal_identity             : identity -> identity -> bool
+  val get_identity               : unit -> identity Lwt.t
+  val max_resource               : int
+  val max_identity_per_resource  : int
+end
+
+module Simple(A : SIMPLE) = Make
+  (struct
```
Comment (vasilisp) [inline src/lib/eliom_notif.server.ml]: Same as above for `SIMPLE` (`ARG_SIMPLE`?). Maybe also rename `Simple` to something that sounds more like a functor, e.g., `Make_simple`?

File: src/lib/eliom_notif.server.ml [hunk 3]
```diff
@@ -188,3 +217,29 @@ module Make (A : S) = struct
...
+
+module type SIMPLE = sig
+  type identity
+  type key
+  type notification
+  val equal_key                  : key -> key -> bool
+  val equal_identity             : identity -> identity -> bool
+  val get_identity               : unit -> identity Lwt.t
+  val max_resource               : int
+  val max_identity_per_resource  : int
```
Comment (vasilisp) [inline src/lib/eliom_notif.server.ml]: Can we simplify this a bit more with sane defaults for `max_*` and maybe even `(=)` for `equal_*`? There is always `Make` for the more involved cases.

File: src/lib/eliom_notif.server.mli [hunk 4]
```diff
@@ -81,11 +74,101 @@ sig
...
       from empty data. *)
   val clean : unit -> unit Lwt.t
 
 end
+
+(** [MAKE] is for making [Make] *)
+module type MAKE = sig
+
+	(** see [S.identity] *)
+  type identity
```
Comment (vasilisp) [inline src/lib/eliom_notif.server.mli]: Indentation :).

File: src/lib/eliom_notif.server.mli [hunk 5]
```diff
@@ -81,11 +74,101 @@ sig
...
+	(** see [S.key] *)
+  type key
+	(** see [S.server_notif] *)
```
Comment (vasilisp) [inline src/lib/eliom_notif.server.mli]: An empty line before this will prevent Ocamldoc ambiguity.

File: src/lib/eliom_notif.server.mli [hunk 6]
```diff
@@ -81,11 +74,101 @@ sig
...
+  (** [equal_identity] is the same as [equal_key] but for values of type
+      [identity].*)
+  val equal_identity             : identity -> identity -> bool
+
+  (** [get_identity] is a function returning a value of type [identity]
+      corresponding to a client. *)
+  val get_identity               : unit -> identity Lwt.t
+
+  (** [max_resource] is the initial size for the hash table storing the data of
+      clients listenning on resources, for best results it should be on the
```
Comment (vasilisp) [inline src/lib/eliom_notif.server.mli]: listening

File: src/lib/eliom_notif.server.mli [hunk 7]
```diff
@@ -81,11 +74,101 @@ sig
...
+      order of the expected number of different resources one may want to be
+      able to listen to. *)
+  val max_resource               : int
+
+  (** [max_identity_per_resource] is the initial size for the tables storing the
+      data of clients listenning on one given resource, fo best results it
```
Comment (vasilisp) [inline src/lib/eliom_notif.server.mli]: listening, as above.

**Comment by jrochel** [comment]
Thanks Vasilis. All implemented. Currently testing.

**Comment by jrochel** [comment]
The code is pretty buggy. I will have to port the current Os_notif code to Eliom_notif.

**Comment by jrochel** [comment]
There is an issue with initialisation of this module. The original version in ocsigen-start does `Os_session.on_start_process set_userchannel` while Eliom_notif does its initialisation during `listen` meaning that functions like `client_ev` will fail (with an uninformative exception) if called before `listen`. If `listen` is called twice `notif_e` is overwritten. @balat intends to reflect/discuss with @vasilisp what the desired behaviour is exactly. Once this is clear I will finish the work on this pull request, but until then this pull request is on hold.

**Comment by jrochel** [comment]
Initialisation has been resolved. Ready to pull (IMO)

**Comment by jrochel** [comment]
It doesn't work. Again on hold...

**Pull request merged**